### ISSUE TITLE
Error updating .htaccess file, not enough permissions?

### ISSUE THREAD

**Issue by mayerthomas** [nextcloud/server#3627]
### Steps to reproduce
1. Having config file without parameter: overwrite.cli.url
2. Doing occ command: sudo -u www-data php occ maintenance:update:htaccess

### Expected behaviour
Updating htaccess-file.
or at least:
Bringing an helpful error that is telling about missing overwrite.cli.url-parameter in config.php

### Actual behaviour
Error 
`Error updating .htaccess file, not enough permissions?`
(see lib/private/Setup.php on lines 420 and following)

### Server configuration
**Operating system**: Debian 8

**Web server:** Apache

**Database:** maria-db

**PHP version:** 7

**Nextcloud version:** 11.0.1

**Updated from an older Nextcloud/ownCloud or fresh install:** tested with updated and fresh installed version

**Where did you install Nextcloud from:** original website

**List of activated apps:** none

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

**Comment by tessus** [comment]
I have the same problem, even if I set .htaccess to 777. 
I'm using nextcloud 11.0.2.

**Comment by nickvergessen** [comment]
Who owns the `.htaccess` file on your server?

**Comment by tessus** [comment]
@nickvergessen not sure, if your question was directed to me, but here's my answer:

it is owned by the user that runs the web server: nobody:nobody 644

But as mentioned before, even if I set it to 777, I get the same error. Although I noticed something else. After reading up on this, it seems that this error message is just wrong. As soon as I added `'overwrite.cli.url'` to the `config.php`, the occ command worked like a charm.

So something is seriously off here. Why would this paramter even have to be set? It changes nothing in the .htaccess output. IMO this is a bug:

- occ should complete without error, no matter if `overwrite.cli.url` is set
- occ should print correct error messages

**Comment by nickvergessen** [comment]
`overwrite.cli.url` is used in the htaccesss.
The value should always be set.

That is also why it errored. When the value is not set, it returns with a failure....
https://github.com/nextcloud/server/blob/3c1809681264626ee556c4cf4a4f51b77462ff3e/lib/private/Setup.php#L414-L417

I will adjust the check in the admin section to point this out correctly.

**Comment by tessus** [comment]
@nickvergessen can we please discuss this for a second. or at least, please try to understand my reasoning:

A parameter called `overwrite.cli.url` should not be required, since it name suggests to override (I think the overwrite was a translation error) a default behavior.
As far as I can ascertain, the value is used for the RewriteBase. The web path should be detected automatically. The only reason for an override parameter would be, if there was a proxy (or any other setup) involved that made the detection unreliable. For standard installations, nobody should have to set this parameter.

If you really need the value in the config file, please rename the option/parameter accordingly.

**Comment by nickvergessen** [comment]
It's also required for link generation in cron jobs and console scripts.
Yes it's a bad name, but renaming config parameters is very awkward because it breaks provisioning scripts and/or makes the code to read the values unnecessary difficult. I'm sorry that you got confused. But we also set this value on install and update, so not sure how you ended up without that value.

**Comment by tessus** [comment]
@nickvergessen yes, I understand that renaming a parameter can be difficult. I was just suggesting that better naming would be nice.
As another suggestion for renaming a parameter you could create an alias from old to new (until you deprecate and remove it at one point).
maybe I ended up without that parameter, because I migrated from oc 8.2.9 -> nc 9.0.57 -> nc 10.0.4 -> 11.0.2.

**Comment by peterpt** [comment]
to fix this error , setup permissions on read/write for www-data group on .htaccess file on nextcloud main server folder .

then apply : 
sudo -u www-data ./occ maintenance:update:htaccess

and it should be fine

**Comment by tessus** [comment]
@peterpt no, this does not fix the problem. the problem was that you got a permission error when `overwrite.cli.url` was missing in the config.

**Comment by peterpt** [comment]
i just had that same error 2 hours ago when i upgraded to latest version and i fixed up that way by allowing www-data group to read/write on .htaccess .

**Comment by tessus** [comment]
Right, because your permissions were wrong in the first place. Mine were not. :-)

The problem was that people got a permission error, even though the permissions were fine.

**Comment by peterpt** [comment]
I see , it looks like it a related problem with a different solution .
I have no idea if in your config you have the database on nextcloud server directory , but 
i first migrated from owncloud 9 to nextcloud 10 , i backed up my config.php and deleted basically my old owncloud directory , then i installed nextcloud 10 on webserver and moved by config.php to config folder on nextcloud server directory .
I did not had to worry about the database because my nextcloud database is not stored on webserver "data" folder .

After that used the owncloud script to set permissions on my webserver folders .
https://doc.owncloud.org/server/9.0/admin_manual/installation/installation_wizard.html#strong-perms-label

All i had to do was to adapt their script to my current server path directory , and set also the location where my real "data" folder was stored .

After this point , was straight ahead and everything worked perfectly .
I use mariadb .

**Comment by kbabioch** [comment]
Same problem here. As soon as I delete `overwrite.cli.url` from the config, I get a permission error. This is seriously wrong, and the error message is deeply misleading!

**Comment by vitalymir** [comment]
Fix the problem:
```
sudo chmod g+w .htaccess
sudo -u www-data php occ maintenance:update:htaccess
sudo chmod g-w .htaccess
```

**Comment by marvinjuette** [comment]
Today I experienced something similar. I just post this in order to provide an workaround for other. I host nextcloud in a docker container. And when I used the bash inside the container to use `sudo -u www-data php -d memory_limit=-1 occ maintenance:update.htaccess` I got the permission or invalid overwrite.cli.url error 

So I tried using the WebOCC Plugin available in the Apps Store and used it to update the htaccess and it worked.
